pyasic/miners/backends/epic.py

```diff
@@ -0,0 +1,297 @@
+                        float(float(hashrate/ 1000000)), 2)
+            except (LookupError, ValueError, TypeError) as e:
+                logger.error(e)
+                pass
+    
+    async def get_nominal_hashrate(self, web_summary: dict = None) -> Optional[float]:
+        # get hr from API
+        if not web_summary:
+            try:
+                web_summary = await self.web.summary()
+            except APIError:
+                pass
+
+        if web_summary:
+            try:
+                hashrate = 0
+                if web_summary["HBs"] != None:
+                    for hb in web_summary["HBs"]:
+                        if hb["Hashrate"][1] == 0:
+                            ideal = 1.0
+                        else:
+                            ideal = hb["Hashrate"][1]/100
+                            
+                        hashrate += hb["Hashrate"][0]/ideal
+                    return round(
+                        float(float(hashrate/ 1000000)), 2)
+            except (IndexError, KeyError, ValueError, TypeError) as e:
+                logger.error(e)
+                pass
+
+
+    async def get_fw_ver(self, web_summary: dict = None) -> Optional[str]:
+        if not web_summary:
+            web_summary = await self.web.summary()
+
+        if web_summary:
+            try:
+                fw_ver = web_summary["Software"]
+                fw_ver = fw_ver.split(" ")[1].replace("v", "")
+                return fw_ver
+            except KeyError:
+                pass
+
+    async def get_fans(self, web_summary: dict = None) -> List[Fan]:
+        if not web_summary:
+            try:
+                web_summary = await self.web.summary()
+            except APIError:
+                pass
+
+        fans = []
+
+        if web_summary:
+            for fan in web_summary["Fans Rpm"]:
+                try:
+                    fans.append(Fan(web_summary["Fans Rpm"][fan]))
+                except (LookupError, ValueError, TypeError):
+                    fans.append(Fan())
+        return fans
+    
+    async def get_hashboards(self, web_summary: dict = None, web_hashrate: dict= None) -> List[HashBoard]:
+        if not web_summary:
+            try:
+                web_summary = await self.web.summary()
+            except APIError:
+                pass
+        if not web_hashrate:
+            try:
+                web_hashrate = await self.web.hashrate()
+            except APIError:
+                pass
+        hb_list = []
+        if web_summary["HBs"] != None:
+            for hb in web_summary["HBs"]:
+                for hr in web_hashrate:
+                    if hr["Index"] == hb["Index"]:
+                        num_of_chips = len(hr["Data"])
+                        hashrate = hb["Hashrate"][0]
+                        hb_list.append(HashBoard(slot=hb["Index"], expected_chips=num_of_chips, missing=False,hashrate=round(hashrate/1000000,2), chips=num_of_chips,temp=hb["Temperature"]))
```

Comment:
I would prefer to move the creation of the `HashBoard`s to the creation of the inital list, and then add to the defaults, as that will account better for missing boards.

E.G. - 
```python3
hb_list = [HashBoard(slot=i, expected_chips=self.nominal_chips) for i in range(self.ideal_hashboards)]
```
Then index in by `hb["index"]`

Reply:
This makes sense, so for our API we will only return if the HB does return so there can be a situation where for examples 0,2 return but no HB 1. I assume the best way to deal with this is to use the missing parameter.

One last question on this, we don't determine models based on NoPic, Pic, instead we use num of chips/EEPROM which wouldn't 100% align with the ideal number of chips being hardcoded. Is it ok if I update that here as I was in the creation of the HashBoard?

eg. S19j 90TH, actually has two versions, 108 chip and 88 chip

Reply:
Any chance you can add that check in as part of the `MinerFactory.get_miner_model_epic` function?  Just check if the model string is `S19j` and then get chip count and choose either the normal or noPIC version?

Technically it shouldn't matter, I would just prefer to have some way to differentiate them at the model level, since they are actually different machines.

Reply:
Let me give this a think. Technically we don't look at all internally if is Pic or noPIC, but let me think if there is something that might give that information. I may even take this back and add it as an API endpoint on our side to keep it cleaner, but that will require a change at a later point

Reply:
Do you want me to wait on this for now, or just merge?  I don't mind merging it now, worst thing this causes is a slightly strange expected chip count on a miner, but any reasonable user can tell that's a normal machine.  Only concern is automated tooling, even in `MinerData` a nominal chip % is calculated, and anything looking at that will assume the miner is bad.

In any case, still a beta version, so if you have plans to change this on the hardware side it can wait.

Reply:
I think it's okay to merge, will work on something ePIC side to improve this. I think something like that would be useful for many other things as well